docs/mp/lra.adoc

```diff
@@ -0,0 +1,460 @@
+the whole architecture is highly scalable.
+
+image::lra/lra-complete-lb.svg[Complete]
+
+In case of error during the LRA transaction, participant reports cancel of LRA to coordinator.
+<<Coordinator, Coordinator>> calls compensate on all the joined participants.
+
+image::lra/lra-compensate-lb-error.svg[Cancel]
+
+When participant joins the LRA with timeout defined `@LRA(value = LRA.Type.REQUIRES_NEW, timeLimit = 5, timeUnit = ChronoUnit.MINUTES)`, coordinator compensate if timeout occurs before close is reported by participants.
+
+image::lra/lra-compensate-lb-timeout.svg[Timeout]
+
+== API
+
+=== Participant
+
+The Participant, or Compensator, is an LRA resource with at least one of the JAX-RS(or non-JAX-RS) methods annotated with
+{javadoc-link}Compensate.html[@Compensate] or {javadoc-link}AfterLRA.html[@AfterLRA].
+
+
+=== @LRA [[lra-method]]
+
+{javadoc-link}ws/rs/LRA.html[~javadoc~]
+
+Marks JAX-RS method which should run in LRA context and needs to be accompanied by at least minimal set of mandatory
+participant methods(<<compensate-participant-method,Compensate>> or <<after-participant-method,AfterLRA>>).
+
+LRA options:
+
+* {javadoc-link}ws/rs/LRA.html#value--[value]
+** {javadoc-link}ws/rs/LRA.Type.html#REQUIRED[REQUIRED] join incoming LRA or create and join new
+** {javadoc-link}ws/rs/LRA.Type.html#REQUIRES_NEW[REQUIRES_NEW] create and join new LRA
+** {javadoc-link}ws/rs/LRA.Type.html#MANDATORY[MANDATORY] join incoming LRA or fail
+** {javadoc-link}ws/rs/LRA.Type.html#SUPPORTS[SUPPORTS] join incoming LRA or continue outside LRA context
+** {javadoc-link}ws/rs/LRA.Type.html#NOT_SUPPORTED[NOT_SUPPORTED] always continue outside LRA context
+** {javadoc-link}ws/rs/LRA.Type.html#NEVER[NEVER] Fail with 412 if executed in LRA context
+** {javadoc-link}ws/rs/LRA.Type.html#NESTED[NESTED] create and join new LRA nested in the incoming LRA context
+* {javadoc-link}ws/rs/LRA.html#timeLimit--[timeLimit] max time limit before LRA gets cancelled automatically by <<Coordinator, coordinator>>
+* {javadoc-link}ws/rs/LRA.html#timeUnit--[timeUnit] time unit if the timeLimit value
+* {javadoc-link}ws/rs/LRA.html#end--[end] when false LRA is not closed after successful method execution
+* {javadoc-link}ws/rs/LRA.html#cancelOn--[cancelOn] which HTTP response codes of the method causes LRA to cancel
+* {javadoc-link}ws/rs/LRA.html#cancelOnFamily--[cancelOnFamily] which family of HTTP response codes causes LRA to cancel
+
+
+Method parameters:
+
+* Header {javadoc-link}ws/rs/LRA.html#LRA_HTTP_CONTEXT_HEADER[LRA_HTTP_CONTEXT_HEADER] - id of the LRA transaction
+
+[source,java]
+----
+@PUT
+@LRA(LRA.Type.REQUIRES_NEW, timeLimit = 500, timeUnit = ChronoUnit.MILLIS)
+@Path("start-example")
+public Response startLra(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId, String data)
+----
+
+=== @Compensate [[compensate-participant-method]]
+
+{javadoc-link}Compensate.html[~javadoc~]
+
+WARNING: Expected to be called by LRA <<Coordinator, coordinator>> only!
```

Comment:
Use Caution to indicate the possibility of damage to a program, system, or data. The caution notice should always precede the step where the risk occurs.

Reply:
Fixed)